src/oom/oomd-util.c

```diff
@@ -280,6 +281,144 @@ int oomd_cgroup_kill(const char *path, bool recurse, bool dry_run) {
         return !set_isempty(pids_killed);
 }
 
+static void sd_oomd_kill_state_dec_pending(struct sd_oomd_kill_state *ks) {
+        if (!ks || ks->refs <= 0)
+                return;
+
+        if (--ks->refs == 0) {
+                if (ks->path) {
+                        oomd_cgroup_kill(ks->path, ks->recurse, ks->prekill_deadline, ks->e);
+                        free(ks->path);
+                }
+                free(ks);
+        }
+}
+
+DEFINE_TRIVIAL_CLEANUP_FUNC_FULL(struct sd_oomd_kill_state*, sd_oomd_kill_state_dec_pending, NULL);
+
+static int prekill_callback(
+                sd_varlink *link,
+                sd_json_variant *parameters,
+                const char *error_id,
+                sd_varlink_reply_flags_t flags,
+                void *userdata) {
+
+        assert(userdata);
+
+        log_debug("oomd prekill hook returned");
+
+        if (error_id)
+                log_warning("oomd prekill hook returned error: %s", error_id);
+
+        sd_varlink_unref(link);
+        sd_oomd_kill_state_dec_pending((struct sd_oomd_kill_state *)userdata);
+
+        return 0;
+}
+
+static int oomd_prekill_hook(const char *path, usec_t prekill_deadline, sd_event *e) {
```

Comment:
I'd do a shortcut when no prekill_hooks. No need to do allocations or anything else.

Reply:
> the response is already logged in prekill_callback()

yes, but with log_debug. IMO it worth log_info. Not a strong opinion though